packages/x-data-grid/src/components/columnSelection/GridHeaderCheckbox.tsx

```diff
@@ -129,11 +129,16 @@ const GridHeaderCheckbox = React.forwardRef<HTMLButtonElement, GridColumnHeaderP
       isChecked ? 'checkboxSelectionUnselectAllRows' : 'checkboxSelectionSelectAllRows',
     );
 
+    const checked =
+      rootProps.indeterminateCheckboxBehavior === 'select'
+        ? isChecked && !isIndeterminate
+        : isChecked;
```

Comment:
Why allow to control this behavior? It feels like only `indeterminateCheckboxBehavior: 'select'` is correct, and it's not a breaking change since the API stays the same, we fix a UX bug.

Reply:
We had a [discussion](https://github.com/mui/mui-x/pull/13757#discussion_r1700587177) on this point where it was suggested that some users might want to keep the existing behavior which is also consistent with some popular applications (like GMail).

So probably allowing the users to control it while having the most desirable behavior (select) by default could be a nice middle ground.

Do you think allowing to control it would make it confusing from a UX pov?

Reply:
> so it wouldn't be a breaking change

👍 Though we have previously considered a few behavioral changes to be breaking since they might break some users' tests.

Reply:
An alternative approach if we don't want a new prop is to support `event.defaultMuiPrevented` on the `onChange` event of the checkbox and add a doc example where the application unselects all items when the checkbox comes from an indeterminate state.

Reply:
My thoughts with asking Why. I saw an opportunity to be lazy, from which we can learn. It could look like this:

- Maybe Gmail is just an engineer who never noticed
- We try to keep the codebase simpler: we inverse the behavior.
- We release it.
- We wait, if we see people complaining, we introduce a prop to inverse the behavior

Reply:
Thank you for expanding. This way of approaching the problem makes sense to me.

Just for reference, I'll share the reason why I assumed the need to retain the possibility of current behavior:
1. #7753 has been open for more than 1.5 years yet we received a few comments and only 5 upvotes. We also have another issue with 6 upvotes, so a combined 11 upvotes. This may imply either of the following
   - The users do not care about this behavior. It doesn't matter if it's Select or Unselect.
   - The users are OK with the current behavior.
   - Some users might deliberately want the current behavior. (This one is the main concerning point for me, if `true`)
2. #3309 also reasons about [allowing](https://github.com/mui/mui-x/issues/3309#issuecomment-1007253147) the customization between both behaviors.
3. Some users already find current behavior intuitive. (Examples: [1](https://github.com/mui/mui-x/issues/3309#issuecomment-1007253147), [2](https://github.com/mui/mui-x/issues/7753#issuecomment-1552764286))

But since point 1 is mainly an assumption that is not validated by many users, I'm OK with starting with reversing the behavior unless there's some disagreement in @mui/xgrid?

Regarding marking this as breaking or not, I'm doubtful. IMO, to be on the safe side, it'd be better to mark it as breaking, why?
1. The v8 alpha is just around the corner.
2. We have treated similar (behavioral) changes breaking in the past.

Unless we are sure to treat it as a bug. I personally consider it a behavioral change that is more aligned with the majority of users. Thoughts?

Reply:
I think this is a breaking change, and I would keep it behind a prop option. Also as a gmail user, I'm used to the gmail behavior regardless of if it's right or wrong, and I doubt there is a right or wrong on this question.

Reply:
Unless we want to do our own research/user testing for this, I would propose we follow the behaviour that we see more commonly across the web and other grids. We want the grid to feel as familiar as possible to the majority of end users.

Below are some implementation examples from the top search results for "mixed-state checkbox" and "indeterminate checkbox", all of which "select all" in an indeterminate state. I haven't found any others (aside from Gmail) that do the opposite.
- https://www.w3.org/WAI/ARIA/apg/patterns/checkbox/examples/checkbox-mixed/
- https://css-tricks.com/indeterminate-checkboxes/
- https://pauljadam.com/demos/mixed-checkbox.html
- https://dequeuniversity.com/library/aria/checkbox-tri
- https://uiplaybook.dev/play/checkbox
- https://www.sitelint.com/blog/definitive-guide-to-indeterminate-state-of-a-checkbox

It is worth pointing out that even Gmail are inconsistent with their selection behaviour. On the Gmail Android app, the default option once you have selected an email is "Select all", the opposite to how it behaves on the web.

<img src="https://github.com/user-attachments/assets/f71ee0ff-5458-44e5-a4d4-daf2fe687b7e" width="300">

---

To avoid breaking user tests, we could introduce the option to reverse the behaviour now and change the default for the next major?

Reply:
@KenanYusuf In v8, could we deprecate the prop at the same time? The maybe this could be an interesting compromise with the more drastic path I proposed.

Reply:
> To avoid breaking user tests, we could introduce the option to reverse the behaviour now and change the default for the next major?

Sounds good to me.

---

Now that I think about it, gmail's behavior sounds more right to me. Which of these cases sounds more likely to you:
 - After selecting a few entries, the user now wants to select all entries
 - After selecting a few entries, the user now wants to deselect all entries
I feel like option 2 is much more useful. The flow of using a selectable grid would be: I select a few rows, I do an operation on them, then I deselect them. I find it much harder to think about a use-case where I want to select all after having selected some.

Reply:
Once the default is changed we could use the telemetry that is being built to track how many people actually want to keep the behavior as it is now

Reply:
> Below are some implementation examples from the top search results for "mixed-state checkbox" and "indeterminate checkbox", all of which "select all" in an indeterminate state. I haven't found any others (aside from Gmail) that do the opposite.

[Angular material example](https://material.angular.io/components/checkbox/overview#checkbox-overview) selects all first as well
Unfortunately, they don't have a [grid selection example](https://material.angular.io/components/table/overview#selection)

Reply:
There are cases for both. At least if we go with the behaviour that is most common, it should behave as expected to a larger audience. The worst case in either scenario is the user has to click an extra time to get to the result they were expecting.

> On @KenanYusuf's proposal, in v8, could we deprecate the prop at the same time?

I don't feel strongly either way. If the prop we introduce is not adding much in terms of complexity and bundle size, I don't see much harm in keeping it if it's helpful to some users.

Reply:
Thanks everyone for your input.
Let's allow the users to customize the behavior now without the breaking change.
About deprecating the prop or keeping the customization, we can continue the discussion in #14300